### PR TITLE
Alloc domain python bug fix

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixing a bug in computing `expanded` through `TensorRecord`;
Adding a python test to verify the failure/fix.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

File: python_tests/test_python_frontend.py [hunk 1]
```diff
@@ -1407,20 +1407,29 @@ def test_stride_order_with_explicit_broadcast(self):
...
 
-            fd.add_output(t2)
+            fd.add_output(t3)
+            fd.add_output(t4)
 
         nvf_out, _ = self.exec_nvfuser(fusion_func, inputs)
         eager_out = inputs[0] + inputs[1]
         self.assertEqual(eager_out, nvf_out[0])
+        eager_out = inputs[2] + 3.0
+        self.assertEqual(eager_out, nvf_out[1])
```
Comment (wujingyue) [inline python_tests/test_python_frontend.py:1432]: Nit: As a convention, `self.assertEqual` expects the actual result on the left and expected on the right. This order makes the error message easier to understand.

File: csrc/python_frontend/fusion_record.h [hunk 2]
```diff
@@ -1229,15 +1229,24 @@ struct TensorRecord : RecordFunctor {
...
+    for (const auto index : c10::irange(rank)) {
       // since contiguity_ vector is given to the corresponding order in alloc
       // domain, while is_expand is given to root domain, we need to map it
       // correctly with `contig_index` and `index`.
-      const auto index = stride_order_.empty()
-          ? contig_index
-          : rank - 1 - static_cast<size_t>(stride_order_[contig_index]);
+      //
+      // stride_order[i] indicates that:
+      //   `root_domain[i]` maps to `alloc_domain[rank - 1 - stride_order_[i]]`
```
Comment (wujingyue) [inline csrc/python_frontend/fusion_record.h:1238]:
```suggestion
      //   `rfactor_domain[i]` (and therefore `root_domain[i]` for input) maps to `alloc_domain[rank - 1 - stride_order_[i]]`
```

File: csrc/python_frontend/fusion_record.h [hunk 3]
```diff
@@ -1229,15 +1229,24 @@ struct TensorRecord : RecordFunctor {
...
+      //
+      // Hence `index` on root domain would be corresponding to the contiguity
+      // index `contig_index = rank - 1 - stride_order[index]`
+      const auto contig_index = stride_order_.empty()
+          ? index
+          : rank - 1 - static_cast<size_t>(stride_order_[index]);
+      bool is_broadcast = !contiguity_[contig_index].has_value();
```
Comment (wujingyue) [inline csrc/python_frontend/fusion_record.h:1245]:
```suggestion
      const bool is_broadcast = !contiguity_[contig_index].has_value();
```

File: csrc/python_frontend/fusion_record.h [hunk 4]
```diff
@@ -1229,15 +1229,24 @@ struct TensorRecord : RecordFunctor {
...
       bool has_symbolic_size = (shape_[index] == -1);
```
Comment (wujingyue) [inline csrc/python_frontend/fusion_record.h:1246]:
```suggestion
      const bool has_symbolic_size = (shape_[index] == -1);
```

File: csrc/python_frontend/fusion_record.h [hunk 5]
```diff
@@ -1229,15 +1229,24 @@ struct TensorRecord : RecordFunctor {
...
+      // A root dimension is expand dimension if:
+      //   The dimension is marked a broadcast; and
+      //   The dimension has a symbolic size (expand extend)
```
Comment (wujingyue) [inline csrc/python_frontend/fusion_record.h:1249]: Wait, is this right? I'm pretty sure I've seen a broadcast dimension with a constant expand extent.

**Review by wujingyue — CHANGES_REQUESTED** [review]

File: csrc/python_frontend/fusion_record.h (hunk 5 above)
Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:1249]: hmmm. That's a good catch... I guess the real case that I need to check is `not broadcast` i.e. `!= -1`?!

Let me try some more tests.

Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:1250]: You caught me there.

Fixed the logic here and added another entry in the python test to verify that.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

File: csrc/python_frontend/fusion_record.h (hunk 5 above)
Comment (wujingyue) [reply csrc/python_frontend/fusion_record.h:1250]: Did you forget to `git push` your latest changes?

**Review by wujingyue — CHANGES_REQUESTED** [review]

**Comment by jjsjann123** [comment]
!build

File: csrc/python_frontend/fusion_record.h (hunk 5 above)
Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:1250]: Oops messed up my history. just pushed.

**Review by jacobhinkle — APPROVED** [review]
This changes the lookup of contiguity to use the inner-first index used in the stride-order convention. So contiguity vector should match allocation domain. LGTM

File: csrc/python_frontend/fusion_record.h [hunk 6]
```diff
@@ -1229,16 +1229,26 @@ struct TensorRecord : RecordFunctor {
...
+      // Hence `index` on root domain would be corresponding to the contiguity
+      // index `contig_index = rank - 1 - stride_order[index]`
+      const auto contig_index = stride_order_.empty()
+          ? index
+          : rank - 1 - static_cast<size_t>(stride_order_[index]);
+      const bool is_broadcast = !contiguity_[contig_index].has_value();
+      const bool has_non_broadcast_size = (shape_[index] != 1);
+      // A root dimension is expand dimension if:
+      //   The dimension is marked a broadcast; and
+      //   The dimension has a symbolic size (expand extend)
```
Comment (wujingyue) [inline csrc/python_frontend/fusion_record.h:1250]:
```suggestion
      //   The dimension has an expanded extent.
```

File: python_tests/test_python_frontend.py [hunk 7]
```diff
@@ -1407,20 +1407,41 @@ def test_stride_order_with_explicit_broadcast(self):
...
+            fd.add_output(t6)
 
         nvf_out, _ = self.exec_nvfuser(fusion_func, inputs)
         eager_out = inputs[0] + inputs[1]
-        self.assertEqual(eager_out, nvf_out[0])
+        self.assertEqual(nvf_out[0], eager_out)
+        eager_out = inputs[2] + 3.0
+        self.assertEqual(nvf_out[1], eager_out)
+        eager_out = inputs[3] * 3.0
+        self.assertEqual(nvf_out[2], eager_out)
```
Comment (wujingyue) [inline python_tests/test_python_frontend.py:1444]:
```suggestion
        self.assertEqual(nvf_out[0], inputs[0] + inputs[1])
        self.assertEqual(nvf_out[1], inputs[2] + 3.0)
        self.assertEqual(nvf_out[2], inputs[3] * 3.0)
```

**Review by wujingyue — APPROVED** [review]

**Pull request merged**